types_test.go

```diff
@@ -2337,3 +3246,179 @@ func TestReceiveSet(t *testing.T) {
+	if conn.conn.protocolVersion.LT(protocolVersion0p11) {
+		t.Skip()
+	}
+
+	ctx := context.Background()
+
+	err := conn.RawTx(ctx, func(ctx context.Context, tx *Tx) error {
+		e := tx.Execute(ctx, `
+			CREATE TYPE Sample {
+				CREATE PROPERTY simple_scalar -> int64;
+				CREATE PROPERTY array -> array<int64>;
+				CREATE PROPERTY tuple -> tuple<int64, int64>;
+				CREATE PROPERTY named_tuple -> tuple<a: int64, b: int64>;
+				CREATE LINK object -> Sample;
+				CREATE MULTI LINK set_ -> Sample;
+			};
+
+			# all fields are missing
+			INSERT Sample;
+		`)
+		if e != nil {
+			return e
+		}
+
+		type Sample struct {
+			SimpleScalar CustomOptionalInt16 `edgedb:"simple_scalar"`
+			Array        []int64             `edgedb:"array"`
+			Tuple        OptionalTuple       `edgedb:"tuple"`
+			NamedTuple   OptionalNamedTuple  `edgedb:"named_tuple"`
+			Object       OtherSample         `edgedb:"object"`
+			Set          []Sample            `edgedb:"set_"`
+		}
+
+		var result Sample
+		e = tx.QueryOne(ctx, `
+			SELECT Sample {
+				simple_scalar,
+				array,
+				tuple,
+				named_tuple,
+				object: { simple_scalar },
+				set_: { simple_scalar },
+			}
+			LIMIT 1`,
+			&result,
+		)
+		if e != nil {
+			return e
+		}
+
+		expected := Sample{
+			SimpleScalar: CustomOptionalInt16{},
+			Array:        nil,
+			Tuple:        OptionalTuple{},
+			NamedTuple:   OptionalNamedTuple{},
+			Object:       OtherSample{},
+			Set:          []Sample{},
+		}
+		assert.Equal(t, expected, result)
+
+		e = tx.QueryOne(ctx, `
+			WITH
+				object := (INSERT Sample { simple_scalar := 2 }),
+				set_ := (INSERT Sample { simple_scalar := 3 }),
+				inserted := (INSERT Sample {
+					simple_scalar := 1,
+					array := [1],
+					tuple := (1, 2),
+					named_tuple := (a := 1, b := 2),
+					object := object,
+					set_ := set_,
+				})
+			SELECT inserted {
+				simple_scalar,
+				array,
+				tuple,
+				named_tuple,
+				object: { simple_scalar },
+				set_: { simple_scalar },
+			}
+			LIMIT 1`,
+			&result,
+		)
+		if e != nil {
+			return e
+		}
+
+		expected = Sample{
+			SimpleScalar: CustomOptionalInt16{
+				value:   1,
+				present: true,
+			},
+			Array: []int64{1},
+			Tuple: OptionalTuple{
+				Zero: int64(1),
+				One:  int64(2),
+			},
+			NamedTuple: OptionalNamedTuple{
+				A:       int64(1),
+				B:       int64(2),
+				present: true,
+			},
+			Object: OtherSample{
+				SimpleScalar: CustomOptionalInt16{
+					value:   2,
+					present: true,
+				},
+			},
+			Set: []Sample{{
+				SimpleScalar: CustomOptionalInt16{
+					value:   3,
+					present: true,
+				},
+			}},
+		}
+		expected.Tuple.SetMissing(false)
+		expected.Object.SetMissing(false)
+
+		assert.Equal(t, expected, result)
+
+		type WrongType struct {
+			SimpleScalar int64 `edgedb:"simple_scalar"`
+		}
+
+		var result2 WrongType
+		e = tx.QueryOne(ctx, `
+			SELECT Sample { simple_scalar } LIMIT 1`,
+			&result2,
+		)
+		assert.EqualError(t, e, "edgedb.UnsupportedFeatureError: "+
```

Comment:
This should be `InvalidArgumentError`.  `UnsupportedFeatureError` is a server error.  Binding-originated errors should use the `ClientError` descendants.

Reply:
changed